### ISSUE TITLE
CSV parsing errors are reported as internal errors

### ISSUE THREAD

**Issue by dmajda** [juttle/juttle#562]
Example:

```
$ cat test.csv 
a,b,c
1,2
$ bin/juttle -e 'read file -file "test.csv" -format "csv"'
Error: internal error RuntimeError: Invalid CSV data: Error: Row length does not match headers { [RuntimeError: internal error RuntimeError: Invalid CSV data: Error: Row length does not match headers]
  name: 'RuntimeError',
  code: 'INTERNAL-ERROR',
  info: 
   { error: 'RuntimeError: Invalid CSV data: Error: Row length does not match headers',
     procName: 'read-file',
     location: { filename: 'main', start: [Object], end: [Object] } } }
└───────────┴───────────┘
```

Errors produced by the CSV parser should be reported as normal Juttle errors.

**Comment by dmajda** [comment]
The underlying cause of this is inappropriate runtime error handling. Instead of morphing this issue into a more general one, I captured the problem in #565. Once that issue is resolved, we should verify its fix also fixes the invalid CSV message and close this issue.

**Comment by rlgomes** [comment]
@dmajda so I'll make this into a CSV-PARSING-ERROR instead ?

**Comment by dmehra** [comment]
As you work on this @rlgomes could you also add line number of the parsing error to the error message? That would help immensely with identifying messed up input.

**Comment by dmajda** [comment]
@rlgomes Well, this wouldn’t fix the underlying issue described in #565, which is the main cause of the ugliness.

Looking at the code, `ReadFile` apparently wraps `CSV-ERROR` (or more generally all parser errors) into `INTERNAL-ERROR`. This wrapping should probably be removed as excessive (at least in cases where the parser produces proper Juttle errors, which it should). But in any case, that is a minor part of this issue.

**Comment by rlgomes** [comment]
https://github.com/juttle/juttle/commit/6d94d901842268194c704008865814a47c58e425 @demmer took care of throwing `INVALID-DATA` instead of `INTERNAL-ERROR`.

@dmehra as for the error reporting indicating line and column I'll open a different issue to track that as the currently used parser doesn't support anything like that https://www.npmjs.com/package/csv-parser but I'll see about finding an alternate parser.

**Comment by rlgomes** [comment]
https://github.com/juttle/juttle/issues/592

**Comment by dmajda** [comment]
Reopening, the error message is still ugly with current master:

```
Error: Invalid CSV data: Error: Row length does not match headers { [RuntimeError: Invalid CSV data: Error: Row length does not match headers]
  name: 'RuntimeError',
  code: 'INVALID-DATA',
  info: 
   { type: 'CSV',
     detail: 'Error: Row length does not match headers',
     procName: 'read-file',
     location: { filename: 'main', start: [Object], end: [Object] } } }
└───────────┴───────────┘
```

It will be ugly at least until #565 is resolved (in other words, this issue depends on it).

**Comment by dmajda** [comment]
Just to clarify, when I talked about “internal error” in the description, I didn’t mean the `INTERNAL-ERROR` code, but the fact that the error is not reported like a normal Juttle error — with a message and code followed by source code excerpt.

**Comment by dmajda** [comment]
This got fixed with #565, the output now looks like this:

```
main:1:1: 
   1:read file -file "test.csv" -format "csv"
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Invalid CSV data: "1,2" at line 1 does not match header line (INVALID-DATA)

┌───────────┬───────────┐
└───────────┴───────────┘
Error: Program terminated with errors
```

It is debatable whether a program which produces an error should also produce output, but I see this as an orthogonal question to this issue. Closing.
